design/ErrorHandling.md

```diff
@@ -7,75 +7,117 @@ There are a number of cases that developers or applications need error handling
 
  - *Debugging*: Getting errors synchronously during development, to break in to the debugger.
  - *Telemetry*: Collecting error logs in deployment, for bug reporting and telemetry.
- - *Recovery*: Recovering from recoverable errors (like out-of-memory on resource creation).
- - *Fallback*: Tearing down the application and falling back, e.g. to WebGL, 2D Canvas, or static content.
+ - *Fallible Allocation*: Recovering from recoverable errors (like out-of-memory on resource creation).
+ - *Fatal Errors*: Handling device/adapter loss, either by restoring WebGPU or by fallback to non-WebGPU content.
 
 Meanwhile, error handling should not make the API clunky to use.
 
 There are several types of WebGPU calls that get their errors handled differently:
 
  - *Creation*: WebGPU Object creation.
- - *Encoding*: Recording of GPU commands in `WebGPUCommandEncoder`.
- - *Operations*: Other API calls, such as `WebGPUQueue.submit`.
+ - *Encoding*: Recording of GPU commands in `GPUCommandEncoder`.
+ - *Operations*: Other API calls, such as `GPUQueue.submit`.
 
 ## *Debugging*: Dev Tools
 
 Implementations should provide a way to enable synchronous validation, for example via a debug shim or via the developer tools.
 The extra overhead needs to be low enough that applications can still run while being debugged.
 
-## *Telemetry* \& *Fallback*: Error Logging
+The behavior is implementation-defined. For example, it may throw an exception, or it may just cause the debugger to pause inside the call that fails.
 
-`WebGPUDevice` keeps a log of errors that happened on each thread (main thread or Web Worker).
-The application "eventually" recieves these errors, asynchronously via a callback.
+## *Fatal Errors*: Lost/Recovered Events
 
-These errors are all delivered through one stream - not directly associated with the objects or operations that produced them.
-They should not be used by applications to recover from expected, recoverable errors.
-Instead, the error log may be used for handling unexpected errors in deployment, for bug reporting and telemetry.
+TODO: distinguish device and adapter loss
+TODO: consider case: unplug discrete and switch to integrated
+TODO: consider case: multiple "apps" on one page using different devices
 
-```
-enum WebGPULogEntryType {
-    "device-lost",
-    "validation-error",
-    "recoverable-out-of-memory",
+```webidl
+enum GPUEventType {
+    "gpudevicelost",
+    "gpudevicerecovered"
 };
 
-interface WebGPULogEntry {
-    readonly attribute WebGPULogEntryType type;
-    readonly attribute any object;
+interface GPUDeviceLostEvent : Event {
+    readonly attribute bool unrecoverable;
```

Comment:
nit: this will lead to double negation `if (!unrecoverable)` change to `recoverable` instead?